src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -56,8 +56,417 @@ private static unsafe void ThrowExceptionEncryptDecryptFail(string fullPath)
             throw Win32Marshal.GetExceptionForWin32Error(errorCode, fullPath);
         }
 
+        private static SafeFileHandle? TryOpenTargetNoAlternativeDataStream(
+            string lpFileName,
+            int dwDesiredAccess,
+            FileShare dwShareMode,
+            FileMode dwCreationDisposition,
+            int dwFlagsAndAttributes,
+            ref Interop.Kernel32.WIN32_FILE_ATTRIBUTE_DATA fileInformation,
+            out string openedFileName)
+        {
+            // Default value for openedFileName.
+            openedFileName = lpFileName;
+
+            // Check the file isn't an alternative data stream
+            if (Path.GetFileName(lpFileName.AsSpan()).Contains(':'))
+            {
+                return null;
+            }
+
+            // Try to open it in a way where we can detect symlinks.
+            SafeFileHandle result = Interop.Kernel32.CreateFile(
+                lpFileName,
+                dwDesiredAccess,
+                dwShareMode,
+                dwCreationDisposition,
+                dwFlagsAndAttributes | Interop.Kernel32.FileOperations.FILE_FLAG_OPEN_REPARSE_POINT
+                );
+
+            // Check we successfully opened it.
+            if (result.IsInvalid)
+            {
+                return null;
+            }
+
+            // Read the file attributes.
+            if (!Interop.Kernel32.GetFileAttributesEx(lpFileName, Interop.Kernel32.GET_FILEEX_INFO_LEVELS.GetFileExInfoStandard, ref fileInformation))
+            {
+                result.Dispose();
+                return null;
+            }
+
+            // Deal with the case of a reparse point.
+            if ((fileInformation.dwFileAttributes & Interop.Kernel32.FileAttributes.FILE_ATTRIBUTE_REPARSE_POINT) != 0)
+            {
+                // Get the path of the target.
+                result.Dispose();
+                //try
+                //{
+                /*    */string? target = GetFinalLinkTarget(lpFileName, false);
```

Comment:
Todo: try/catch around this? Maybe even a version which won't throw if we care that much.